### ISSUE TITLE
BatchKVCache: mx.array += in-place mutation corrupts RoPE offsets in batch inference

### ISSUE THREAD

**Issue by nicdavidson** [ml-explore/mlx-lm#1097]
Found this while getting Gemma 4 31B running through batch inference (oMLX server). Single-prompt generation works fine, batch serving produces garbage. Root cause is mx.array += in-place mutation on cache offsets.

The problem

In the batch generation path, the code captures the cache offset before calling update_and_fetch:

```
offset = cache.offset          # reference to mx.array
keys, values = cache.update_and_fetch(keys, values)
# internally: self.offset += keys.shape[2]   mutates in place
queries = rope(queries, offset=offset)  # offset is now WRONG
```

Python ints are immutable: += creates a new object. mx.array is mutable += modifies in place. So offset ends up pointing at the post-update value instead of the pre-update value. RoPE positions are wrong, attention patterns break, output is garbage.

Regular KVCache uses a Python int for offset, so single-prompt inference is unaffected. BatchKVCache uses mx.array, so this only shows up in batch/server mode.

Fix

`offset = cache.offset + 0   # force a copy`

The + 0 allocates a new array with the current value. When update_and_fetch mutates cache.offset, the local copy is untouched.

Possibly related issues

This might explain or contribute to several open batch inference issues:

#965 — KV cache cross-contamination between concurrent requests

#939 — SuScaledRoPE mutates x parameter (same bug class, already acknowledged)

#983, #1035 — BatchRotatingKVCache merge crashes with concurrent requests

PR #999 addresses physical vs logical offset for RotatingKVCache but doesn't cover BatchKVCache.

Repro

Any model served through batch inference with BatchKVCache. I hit it on Gemma 4 31B but it's architecture-independent — the bug is in the cache/generation path, not the model.

**Comment by khsd6327** [comment]
Tested this on oMLX 0.3.2 + mlx-lm 0.31.2 (commit 564281f) with Gemma 4 31B and 26B.

**The `+ 0` fix alone does not resolve the issue for Gemma 4.**

Applied `offset = cache.offset + 0` in `mlx_vlm/models/gemma4/language.py` (lines 207, 210) — batch corruption persists at the same rate (~70-80% failure with 5 concurrent requests).

Also tried:
- Replacing all `self.offset +=` / `self.left_padding -=` in-place mutations in `cache.py` with `self.x = self.x + y` — no effect
- Updating `cache.py` to include the `BatchRotatingKVCache.merge()` fix from PR #1072 — no effect
- Combinations of the above — no effect

**This appears to be Gemma 4 specific, not architecture-independent:**

| Model | Concurrent 5x | Result |
|---|---|---|
| Gemma 4 31B (4bit) | 2/10 OK | Broken — repetition loops, `finish_reason: length` |
| Gemma 4 26B (4bit) | 1/5 OK | Broken — same symptoms |
| Qwen 3.5 122B | 5/5 OK | Clean |
| Qwen 3.5 35B | 5/5 OK | Clean |
| Qwen 3 Coder Next | 5/5 OK | Clean |
| gpt-oss 20B | 5/5 OK | Clean |
| Nemotron 3 Super 120B | 5/5 OK | Clean |

Sequential requests to Gemma 4 work 10/10. The corruption only occurs during batched decode.

Gemma 4's distinguishing feature is hybrid attention (5 sliding_attention + 1 full_attention, `sliding_window=1024`), which means `BatchRotatingKVCache` + `BatchKVCache` are mixed in the same cache list. The other models tested use uniform cache types. This suggests the root cause may be in how the hybrid cache interacts with batched generation, rather than the offset mutation alone.

Environment: macOS, Apple M5 Max 128GB, MLX 0.31.1, mlx-vlm 0.4.3

**Comment by khsd6327** [comment]
**Confirmed: reproducible with pure mlx-lm BatchGenerator, no oMLX involved.**

Loaded Gemma 4 31B via mlx-vlm, extracted `language_model`, patched `LanguageModelOutput` → raw logits, and ran through mlx-lm's `BatchGenerator` directly:

```python
gen = BatchGenerator(lang_model, stop_tokens={1, 50, 106})
gen.insert(prompts=[tokens]*5, max_tokens=[300]*5)
```

Result: **5/5 broken** — all outputs are `SSSSSSSSBBBBBBBB...` (uniform garbage).

This rules out oMLX's scheduler, VLMModelAdapter, _IntOffsetCacheProxy, and all server-side code. The bug is entirely within mlx-lm's `BatchGenerator` + Gemma 4's hybrid cache (`RotatingKVCache` + `KVCache` mixed per layer).

**Comment by angeloskath** [comment]
This is by design. The offset should be used before `update_and_fetch` a bit brittle but not a bug in the cache. Also no need to do `+ 0` which actually runs a kernel adding a 0. Any of `mx.array(x)` or `x[...]` are noops that grab a shallow copy but ensure the data is properly kept unchanged.

If you find any model that is using the offset after `update_and_fetch` then please file that as a bug.
